pennylane/compiler/python_compiler/xdsl_extras/constraints.py

```diff
@@ -173,3 +182,53 @@ class MemRefConstraint(ContainerConstraint):
     @property
     def expected_type(self):
         return MemRefType
+
+
+@dataclass(frozen=True)
```

Comment:
I think if you add a constructor to a dataclass you need to do the following
```suggestion
@dataclass(frozen=True, init=False)
```